superset-frontend/src/middleware/asyncEvent.ts

```diff
@@ -0,0 +1,188 @@
+  parseErrorJson,
+} from '../utils/getClientErrorObject';
+
+export type AsyncEvent = {
+  id: string;
+  channel_id: string;
+  job_id: string;
+  user_id: string;
+  status: string;
+  errors: SupersetError[];
+  result_url: string;
+};
+
+type AsyncEventOptions = {
+  getPendingComponents: (state: any) => any[];
+  successAction: (componentId: number, componentData: any) => { type: string };
+  errorAction: (componentId: number, response: any) => { type: string };
+  processEventsCallback?: (events: AsyncEvent[]) => void; // this is currently used only for tests
+};
+
+type CachedDataResponse = {
+  componentId: number;
+  status: string;
+  data: any;
+};
+
+const initAsyncEvents = (options: AsyncEventOptions) => {
+  const POLLING_DELAY = 250;
+  const {
+    getPendingComponents,
+    successAction,
+    errorAction,
+    processEventsCallback,
+  } = options;
+
+  const middleware: Middleware = <S>(store: MiddlewareAPI<S>) => (
+    next: Dispatch<S>,
+  ) => {
+    const JOB_STATUS = {
+      PENDING: 'pending',
+      RUNNING: 'running',
+      ERROR: 'error',
+      DONE: 'done',
+    };
+    const LOCALSTORAGE_KEY = 'last_async_event_id';
+    const POLLING_URL = '/api/v1/async_event/';
+    let lastReceivedEventId: string | null;
+
+    try {
+      lastReceivedEventId = localStorage.getItem(LOCALSTORAGE_KEY);
+    } catch (err) {
+      console.warn('failed to fetch last event Id from localStorage');
+    }
+
+    const fetchEvents = async (
+      lastEventId: string | null,
+    ): Promise<AsyncEvent[]> => {
+      const url = lastEventId
+        ? `${POLLING_URL}?last_id=${lastEventId}`
+        : POLLING_URL;
+      const { json } = await SupersetClient.get({
+        endpoint: url,
+      });
+
+      return json.result;
+    };
+
+    const fetchCachedData = async (
+      asyncEvent: AsyncEvent,
+      componentId: number,
+    ): Promise<CachedDataResponse> => {
+      let status = 'success';
+      let data;
+      try {
+        const { json } = await SupersetClient.get({
+          endpoint: asyncEvent.result_url,
+        });
+        data = 'result' in json ? json.result[0] : json;
+      } catch (response) {
+        status = 'error';
+        data = await getClientErrorObject(response);
+      }
+
+      return { componentId, status, data };
+    };
+
+    const setLastId = (asyncEvent: AsyncEvent) => {
+      lastReceivedEventId = asyncEvent.id;
+      try {
+        localStorage.setItem(LOCALSTORAGE_KEY, lastReceivedEventId as string);
+      } catch (err) {
+        console.warn('Error saving event ID to localStorage', err);
+      }
+    };
+
+    const processEvents = async () => {
+      const state = store.getState();
+      const queuedComponents = getPendingComponents(state);
+      let events: AsyncEvent[] = [];
+      if (queuedComponents && queuedComponents.length) {
+        try {
+          events = await fetchEvents(lastReceivedEventId);
+          if (events && events.length) {
```

Comment:
Nit: could use an early return here to reduce indentation.

Reply:
That's how I had it initially, but didn't want to duplicate the `processEventsCallback` logic, which is (unfortunately) required to make the tests work properly.